aspnetcore/data/ef-rp/read-related-data.md

```diff
@@ -199,53 +235,106 @@ You've made the following changes to the existing code:
+
+![Instructor-to-Courses m:M](complex-data-model/_static/courseassignment.png)
+
+The `Single` method is used on a collection when the collection has only one item. The `Single` method throws an exception if the collection is empty or if there's more than one item. An alternative is `SingleOrDefault`, which returns a default value (null in this case) if the collection is empty. Using `SingleOrDefault` on an empty collection:
+
+* Results in an exception (from trying to find a `Courses` property on a null reference).
+* The exception message would less clearly indicate the cause of the problem.
+
+The following code populates the view model's `Enrollments` property when a course is selected:
+
+[!code-csharp[Main](intro/samples/cu/Pages/Instructors/Index2.cshtml.cs?name=snippet_courseID)]
+
+Add the following markup to the end of the *Pages/Courses/Index.cshtml* Razor Page:
+
+[!code-cshtml[](intro/samples/cu/Pages/Instructors/IndexRRD.cshtml?range=60-102&highlight=7-)]
+
+The preceding markup displays a list of courses related to an instructor when an instructor is selected.
+
+Test the app. Click on a **Select** link on the instructors page.
 
 ![Instructors Index page instructor selected](read-related-data/_static/instructors-index-instructor-selected.png)
 
-After the code block you just added, add the following code. This displays a list of the students who are enrolled in a course when that course is selected.
+### Show student data
+
+In this section, the app is updated to show the student data for a selected course.
+
+Update the query in the `OnGetAsync` method in *Pages/Instructors/Index.cshtml.cs* with the following code:
+
+[!code-csharp[Main](intro/samples/cu/Pages/Instructors/Index.cshtml.cs?name=snippet_ThenInclude&highlight=6-9)]
+
+Update *Pages/Instructors/Index.cshtml*. Add the following markup to the end of the file:
 
-[!code-html[](intro/samples/cu/Views/Instructors/Index1.cshtml?range=103-125)]
+[!code-cshtml[](intro/samples/cu/Pages/Instructors/IndexRRD.cshtml?range=103-)]
 
-This code reads the Enrollments property of the view model in order to display a list of students enrolled in the course.
+The preceding markup displays a list of the students who are enrolled in the selected course.
 
-Refresh the page again and select an instructor. Then select a course to see the list of enrolled students and their grades.
+Refresh the page and select an instructor. Select a course to see the list of enrolled students and their grades.
 
 ![Instructors Index page instructor and course selected](read-related-data/_static/instructors-index.png)
 
+## Using Single
+
+The `Single` method can pass in the `Where` condition instead of calling the `Where` method separately:
+
+[!code-csharp[Main](intro/samples/cu/Pages/Instructors/IndexSingle.cshtml.cs?name=snippet_single&highlight=21,28-29)]
+
+The preceding `Single` approach provides no benefits over using `Where`. Some developers prefer the `Single` approach style.
```

Comment:
This is causing the whole list to be loading into memory and then the Single call is executing in-memory (with LINQ to Objects) against that list. Is there are reason to do this rather than using SingleAsync on the query itself so that the selection can happen in the database?

Reply:
@ajcvickers  see [code here](https://github.com/aspnet/Docs/blob/e0f83039d06bfec8662656e92c1355f5c31d3d03/aspnetcore/data/ef-rp/intro/samples/cu/Pages/Instructors/IndexSingle.cshtml.cs)

Yes, need to load entire list. This is the index page for instructors. The other sub-queries are if you have selected an instructor on the page.